eax/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Encryption key
+    key: Key<Cipher>,
+}
+
+impl<Cipher> NewAead for Eax<Cipher>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    type KeySize = Cipher::KeySize;
+
+    fn new(key: &Key<Cipher>) -> Self {
+        Self { key: key.clone() }
+    }
+}
+
+impl<Cipher> AeadInPlace for Eax<Cipher>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    type NonceSize = Cipher::BlockSize;
+    type TagSize = <Cmac<Cipher> as Mac>::OutputSize;
+    type CiphertextOverhead = U0;
+
+    fn encrypt_in_place_detached(
+        &self,
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+    ) -> Result<Tag, Error> {
+        if buffer.len() as u64 > P_MAX || associated_data.len() as u64 > A_MAX {
+            return Err(Error);
+        }
+
+        // https://crypto.stackexchange.com/questions/26948/eax-cipher-mode-with-nonce-equal-header
+        // has an explanation of eax.
+
+        // l = block cipher size = 128 (for AES-128) = 16 byte
+        // 1. n ← OMAC(0 || Nonce)
+        // (the 0 means the number zero in l bits)
+        let n = Self::cmac_with_iv(&self.key, 0, nonce);
+
+        // 2. h ← OMAC(1 || associated data)
+        let h = Self::cmac_with_iv(&self.key, 1, associated_data);
+
+        // 3. enc ← CTR(M) using n as iv
+        let mut cipher = ctr::Ctr128::<Cipher>::from_block_cipher(Cipher::new(&self.key), &n);
+        cipher.apply_keystream(buffer);
+
+        // 4. c ← OMAC(2 || enc)
+        let c = Self::cmac_with_iv(&self.key, 2, buffer);
+
+        // 5. tag ← n ^ h ^ c
+        // (^ means xor)
+        Ok(n.zip(h, |a, b| a ^ b).zip(c, |a, b| a ^ b))
+    }
+
+    fn decrypt_in_place_detached(
+        &self,
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+        tag: &Tag,
+    ) -> Result<(), Error> {
+        if buffer.len() as u64 > C_MAX || associated_data.len() as u64 > A_MAX {
+            return Err(Error);
+        }
+
+        // 1. n ← OMAC(0 || Nonce)
+        let n = Self::cmac_with_iv(&self.key, 0, nonce);
+
+        // 2. h ← OMAC(1 || associated data)
+        let h = Self::cmac_with_iv(&self.key, 1, associated_data);
+
+        // 4. c ← OMAC(2 || enc)
+        let c = Self::cmac_with_iv(&self.key, 2, buffer);
+
+        // 5. tag ← n ^ h ^ c
+        // (^ means xor)
+        let expected_tag = n.zip(h, |a, b| a ^ b).zip(c, |a, b| a ^ b);
+
+        let expected_tag = &expected_tag[..tag.len()];
+
+        // Check mac using secure comparison
+        use subtle::ConstantTimeEq;
+        if expected_tag.ct_eq(tag).unwrap_u8() == 1 {
+            // Decrypt
+            let mut cipher = ctr::Ctr128::<Cipher>::from_block_cipher(Cipher::new(&self.key), &n);
```

Comment:
Note that this approach will result in a lower performance for AES compared to types from `aesni`, but I guess there is no good way to solve it right now...

Also does the paper specify counter width? Because implementation in the `ctr` crate uses 64-bit counters.